### PR TITLE
Support mutation protocol for flat mappings with binding transformations

### PR DESCRIPTION
**Pull request by horbe**
#### Which issue(s) this PR fixes:

Fixes #

### CODE REVIEW

File: legend-engine-executionPlan-execution/src/main/java/org/finos/legend/engine/plan/execution/nodes/ExecutionNodeExecutor.java [hunk 1]
```diff
@@ -230,6 +236,26 @@ public Result visit(AllocationExecutionNode allocationExecutionNode)
         {
             result = new ConstantResult(((Map<?, ?>) ((ConstantResult) result).getValue()).get("value"));
         }
+        // TODO HSO: Right place for this? Other streaming result types?
```
Comment (rafaelbey) [inline legend-engine-executionPlan-execution/src/main/java/org/finos/legend/engine/plan/execution/nodes/ExecutionNodeExecutor.java]: I think each result set should implement (or throw if not supported) the `realizeInMemory` method.  Then, this code will be encapsulated on the right place.

One consideration tho is the limit we might want to set when realizing in memory like we do for relational results.

File: legend-engine-pure-code-compiled-core/src/main/resources/core/pure/binding/executionPlan/executionPlan_generation.pure [hunk 2]
```diff
@@ -51,6 +51,7 @@ function meta::external::shared::format::executionPlan::processValueSpecificatio
       f:SimpleFunctionExpression[1]         | $f->getFunctionProcessor($extensions)->eval($f, $state, $extensions, $debug), 
       c:ClusteredValueSpecification[1]      | $c->plan($state.inScopeVars, $state.exeCtx, $extensions, $debug),
       e:ExtendedRoutedValueSpecification[1] | $e.value->processValueSpecification($state, $extensions, $debug),
+      var:VariableExpression[1]             | ^VariableResolutionExecutionNode(varName=$var.name, resultType=^ResultType(type=Any)),
```
Comment (rafaelbey) [inline legend-engine-pure-code-compiled-core/src/main/resources/core/pure/binding/executionPlan/executionPlan_generation.pure]: How hard will be to use the type of the variable expression to set the node type?  Same for the node multiplicity.

File: legend-engine-pure-code-compiled-core/src/main/resources/core/pure/mutation/mutation.pure [hunk 3]
```diff
@@ -0,0 +1,16 @@
+import meta::pure::mapping::*;
+import meta::pure::runtime::*;
+import meta::pure::graphFetch::*;
+
+function meta::pure::mutation::execution::save<T>(type: Class<T>[1], rootGraphFetchTree: RootGraphFetchTree<T>[1], values: T[*], mapping: Mapping[1], runtime: Runtime[1]): T[*]
```
Comment (rafaelbey) [inline legend-engine-pure-code-compiled-core/src/main/resources/core/pure/mutation/mutation.pure]: We should add `<<functionType.NotImplementedFunction>>` to this function.

File: legend-engine-pure-code-compiled-core/src/main/resources/core/pure/protocol/vX_X_X/invocations/execution.pure [hunk 4]
```diff
@@ -184,14 +184,14 @@ function meta::protocols::pure::vX_X_X::invocation::execution::execute::legendEx
...
-                                                    |meta::pure::executionPlan::executionPlan($f, $m, $pureRuntime, if($context->isEmpty(), |^ExecutionContext(), |$context->toOne()), $extensions, noDebug()););
+                                                    |meta::pure::executionPlan::executionPlan($f, if($context->isEmpty(), |^ExecutionContext(), |$context->toOne()), $extensions, debug());,
+                                                    |meta::pure::executionPlan::executionPlan($f, $m, $pureRuntime, if($context->isEmpty(), |^ExecutionContext(), |$context->toOne()), $extensions, debug()););
                                   let boundPlan = meta::pure::executionPlan::platformBinding::localBinding::bindTestPlanToPlatformLocally($basicPlan, $extensions);
-                                  let unused    = if(isOptionSet('ShowLocalPlan'), |println($boundPlan->meta::pure::executionPlan::toString::planToString(true, $extensions)), |[]);
+                                  let unused    = if(true || isOptionSet('ShowLocalPlan'), |println($boundPlan->meta::pure::executionPlan::toString::planToString(true, $extensions)), |[]);
```
Comment (rafaelbey) [inline legend-engine-pure-code-compiled-core/src/main/resources/core/pure/protocol/vX_X_X/invocations/execution.pure]: Cclean up these...

File: legend-engine-pure-code-compiled-core/src/main/resources/core/pure/mutation/mutation_routing.pure [hunk 5]
```diff
@@ -0,0 +1,11 @@
+import meta::pure::graphFetch::*;
+import meta::pure::router::store::metamodel::*;
+import meta::pure::router::store::metamodel::clustering::*;
+
+Class meta::pure::router::store::metamodel::WriteRoutedValueSpecification extends StoreMappingRoutedValueSpecification
```
Comment (rafaelbey) [inline legend-engine-pure-code-compiled-core/src/main/resources/core/pure/mutation/mutation_routing.pure]: As discussed, we need to understand why this is required.  Seems odd

File: legend-engine-xt-relationalStore-grammar/src/main/java/org/finos/legend/engine/language/pure/compiler/toPureGraph/RelationalCompilerExtension.java [hunk 6]
```diff
@@ -483,6 +483,10 @@ public List<Function<Handlers, List<FunctionExpressionBuilderRegistrationInfo>>>
                         ),
                         new FunctionExpressionBuilderRegistrationInfo(Lists.mutable.with(3),
                                 handlers.m(handlers.h("meta::pure::functions::asserts::assertEq_Any_1__Any_1__Function_1__Boolean_1_", false, ps -> handlers.res("Boolean", "one"), ps -> ps.size() == 3 && !handlers.typeOne(ps.get(2), "String")))
+                        ),
+                        // TODO HSO: This is wrong
```
Comment (rafaelbey) [inline legend-engine-xt-relationalStore-grammar/src/main/java/org/finos/legend/engine/language/pure/compiler/toPureGraph/RelationalCompilerExtension.java]: this does not seem wrong, the location (relational extension) is what makes it odd to me.

File: legend-engine-xt-relationalStore-pure/src/main/resources/core_relational/relational/contract/storeContract.pure [hunk 7]
```diff
@@ -166,6 +175,8 @@ function meta::relational::contract::supports(f:FunctionExpression[1]):Boolean[1
             meta::pure::functions::math::sum_Float_MANY__Float_1_,
             meta::pure::functions::math::sum_Integer_MANY__Integer_1_,
             meta::pure::functions::math::sum_Number_MANY__Number_1_,
+            // TODO HSO: Where does this belong?
+            meta::pure::mutation::execution::save_Class_1__RootGraphFetchTree_1__T_MANY__Mapping_1__Runtime_1__T_MANY_,
```
Comment (rafaelbey) [inline legend-engine-xt-relationalStore-pure/src/main/resources/core_relational/relational/contract/storeContract.pure]: I think this belong to any store that wants to support save, hence this seems correct to me.

File: legend-engine-pure-code-compiled-core/src/main/resources/core/pure/mutation/mutation.pure (hunk 3 above)
Comment (horbe) [reply legend-engine-pure-code-compiled-core/src/main/resources/core/pure/mutation/mutation.pure]: Thanks. Done

File: legend-engine-pure-code-compiled-core/src/main/resources/core/pure/protocol/vX_X_X/invocations/execution.pure (hunk 4 above)
Comment (horbe) [reply legend-engine-pure-code-compiled-core/src/main/resources/core/pure/protocol/vX_X_X/invocations/execution.pure]: Done.

File: legend-engine-pure-code-compiled-core/src/main/resources/core/pure/binding/executionPlan/executionPlan_generation.pure (hunk 2 above)
Comment (horbe) [reply legend-engine-pure-code-compiled-core/src/main/resources/core/pure/binding/executionPlan/executionPlan_generation.pure]: This is actual no longer need. Have removed.

File: legend-engine-xt-relationalStore-pure/src/main/resources/core_relational/relational/contract/storeContract.pure (hunk 7 above)
Comment (horbe) [reply legend-engine-xt-relationalStore-pure/src/main/resources/core_relational/relational/contract/storeContract.pure]: Moved into the getSupportedFunctions block

**Pull request closed without merging**